x/mongo/driver/operation/hello.go

```diff
@@ -113,6 +121,430 @@ func (h *Hello) Result(addr address.Address) description.Server {
+	}
+
+	return ""
+}
+
+// appendStringElement appends the key and value to dst. maxLen is the maximum
+// number of bytes that can be appended to dst.
+func appendStringElement(dst []byte, key, value string, maxLen int32) (int32, []byte) {
+	originalLen := int32(len(dst))
+	if int32(len(key)+len(value))+stringElementSize > maxLen {
+		return 0, dst
+	}
+
+	dst = bsoncore.AppendStringElement(dst, key, value)
+
+	bytesWritten := int32(len(dst)) - originalLen
+	return bytesWritten, dst
+}
+
+// appendInt32Element appends the key and value to dst. maxLen is the maximum
+// number of bytes that can be appended to dst.
+func appendInt32Element(dst []byte, key string, value int32, maxLen int32) (int32, []byte) {
+	originalLen := int32(len(dst))
+	if int32(len(key))+int32ElementSize > maxLen {
+		return 0, dst
+	}
+
+	dst = bsoncore.AppendInt32Element(dst, key, value)
+
+	bytesWritten := int32(len(dst)) - originalLen
+	return bytesWritten, dst
+}
+
+// appendClientAppName appends the application name to dst. It builds the
+// application sub-document key-by-key, checking the length of the document
+// after each key. If the document exceeds the maximum length, the key is
+// omitted and the document is ended and returned. MaxLen is the maximum number
+// of bytes that can be appended to dst.
+func appendClientAppName(dst []byte, name string, maxLen int32) (int32, []byte, error) {
+	const key = "application"
+
+	// Validate input parameters.
+	if dst == nil || maxLen <= 0 {
+		return 0, dst, nil
+	}
+
+	// If the size of the embedded document with the "application" key is
+	// greater than the maximum length, then the document is too long and
+	// we should terminate.
+	if embeddedDocumentSize+int32(len(key)) > maxLen {
+		return 0, dst, nil
+	}
+
+	// Account for the null byte at the end of the document.
+	maxLen--
+
+	// Save the original length of dst so we can calculate the number of
+	// bytes written.
+	originalLen := len(dst)
+	idx, dst := bsoncore.AppendDocumentElementStart(dst, key)
+
+	// Decrease the maximum length to account for the bytes appended so far.
+	maxLen -= int32(len(dst) - originalLen)
+
+	_, dst = appendStringElement(dst, "name", name, maxLen)
+
+	var err error
+
+	dst, err = bsoncore.AppendDocumentEnd(dst, idx)
+	if err != nil {
+		return 0, dst, err
+	}
+
+	bytesWritten := int32(len(dst) - originalLen)
+	return bytesWritten, dst, nil
+}
+
+// appendClientDriver appends the driver metadata to dst. It builds the
+// driver sub-document key-by-key, checking the length of the document after
+// each key. If the document exceeds the maximum length, the driver is omitted
+// and the document is ended and returned. MaxLen is the maximum number of
+// bytes that can be appended to dst.
+func appendClientDriver(dst []byte, maxLen int32) (int32, []byte, error) {
+	const key = "driver"
+
+	// Validate input parameters.
+	if dst == nil || maxLen <= 0 {
+		return 0, dst, nil
+	}
+
+	if embeddedDocumentSize+int32(len(key)) > maxLen {
```

Comment:
I've created a fallback for each of those append functions. I'm not sure I follow on how to use ReadLength/UpdateLength to create the truncation. IIUC the first byte in the "dst" will be the length but I'm not sure that actually gets updated until the document is ended. Since all of these apened functions are specific to the client metadata, and they are all called before the client metadata document is ended, the first byte will always be 0 within this context.

This is a really good idea, though. Can you re-review and see if the proposed solution is acceptable?